src/coprocessor/dag/batch/executors/limit.rs

```diff
@@ -0,0 +1,304 @@
+
+            let field_types = vec![
+                field_type(FieldTypeTp::LongLong),
+                field_type(FieldTypeTp::LongLong),
+                field_type(FieldTypeTp::Double),
+            ];
+
+            MockExecutor {
+                data: expect_rows,
+                field_types,
+                offset: 0,
+                cfg: EvalConfig::default(),
+                zero_rows_for_next_batch: false,
+            }
+        }
+    }
+
+    impl BatchExecutor for MockExecutor {
+        #[inline]
+        fn next_batch(&mut self, expect_rows: usize) -> BatchExecuteResult {
+            let upper = if self.zero_rows_for_next_batch {
+                self.zero_rows_for_next_batch = false;
+                // return zero rows
+                self.offset
+            } else if expect_rows + self.offset > self.data.len() {
+                self.data.len()
+            } else {
+                expect_rows + self.offset
+            };
+            let mut pks = VectorValue::with_capacity(self.data.len(), EvalType::Int);
+            let mut foos = VectorValue::with_capacity(self.data.len(), EvalType::Int);
+            let mut bars = VectorValue::with_capacity(self.data.len(), EvalType::Real);
+            for id in self.offset..upper {
+                let (handle, foo, bar) = self.data[id];
+                pks.push_int(Some(handle));
+                foos.push_int(foo);
+                bars.push_real(bar);
+            }
+            self.offset = upper;
+            BatchExecuteResult {
+                data: LazyBatchColumnVec::from(vec![
+                    LazyBatchColumn::Decoded(pks),
+                    LazyBatchColumn::Decoded(foos),
+                    LazyBatchColumn::Decoded(bars),
+                ]),
+                warnings: self.cfg.new_eval_warnings(),
+                is_drained: Ok(self.offset == self.data.len()),
+            }
+        }
+
+        #[inline]
+        fn collect_statistics(&mut self, _: &mut BatchExecuteStatistics) {}
+
+        #[inline]
+        fn schema(&self) -> &[FieldType] {
+            self.field_types.as_ref()
+        }
+    }
+
+    #[test]
+    fn test_limit_normal() {
+        let data = vec![
+            //limit, expect_rows(real get rows)
+            (4, 3), // get less than limit
+            (3, 4), // get more than limit
+            (3, 3), // get equals to limit
+        ];
+        for (limit, get_rows) in data {
+            let src = MockExecutor::new();
+            let mut executor =
+                BatchLimitExecutor::new(src, limit, ExecSummaryCollectorDisabled).unwrap();
+            let res = executor.next_batch(get_rows);
+            if limit <= get_rows {
+                // is drained
+                assert!(res.is_drained.as_ref().unwrap());
+                assert_eq!(res.data.rows_len(), limit);
+            } else {
+                assert!(!res.is_drained.as_ref().unwrap());
+                assert_eq!(res.data.rows_len(), get_rows);
```

Comment:
This test only checks the first call to `next_batch()`. We also need to check for the next call to `next_batch()` until either:
1. the limit is hit
2. the underlying data is drained